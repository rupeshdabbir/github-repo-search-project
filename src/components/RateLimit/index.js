// @flow

import React from 'react';
import {
  StyleSheet,
  css,
} from 'aphrodite/no-important';

type Props = {
  label: string,
  reset: string,
  remaining: number,
  limit: number,
};

export function calculateRemainingColor(remaining: number, limit: number): string {
  if (remaining <= (limit / 4) * 1) {
    return 'RateLimit_low';
  }
  if (remaining <= (limit / 4) * 3) {
    return 'RateLimit_medium';
  }
  return 'RateLimit_high';
}

export default function RateLimit({label, remaining, limit, reset}: Props) {
  const styleKey = calculateRemainingColor(remaining, limit);
  return (
    <p
      className={`${css(styles.RateLimit)} u-flex`}
      title={`Limit resets at ${reset}`}
    >
      <span className={css(styles.RateLimit_label)}>{label}:</span>
      <span className={css(styles[styleKey])}>{remaining}</span>
    </p>
  );
}

RateLimit.defaultProps = {
  remaining: 0,
  limit: 0,
  reset: '',
};

const styles = StyleSheet.create({
  RateLimit: {
    fontSize: 14,
  },

  RateLimit_label: {
    marginRight: 3,
  },

  RateLimit_high: {
    color: '#10ce65',
  },

  RateLimit_medium: {
    color: '#f5f758',
  },

  RateLimit_low: {
    color: '#d6495d',
  },
});
