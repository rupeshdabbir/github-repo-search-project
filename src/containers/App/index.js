// @flow

import React from 'react';
import { HashRouter as Router, Redirect, Switch } from 'react-router-dom';
import SearchScreen from 'screens/Search';
import DefaultLayout from 'screens/Default';

import 'css/global.css';

export default function App() {
  return (
    <Router>
      <Switch>
        <DefaultLayout exact={true} path="/search" component={SearchScreen} />
        <Redirect to="/search" />
      </Switch>
    </Router>
  );
}
