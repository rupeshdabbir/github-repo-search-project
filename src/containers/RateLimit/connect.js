// @flow

import { connect } from 'react-redux';
import get from 'lodash/fp/get';
import format from 'date-fns/format';

function formatDate(timestamp: number): string | void {
  if (!timestamp) {
    return undefined;
  }
  return format(new Date(timestamp * 1000), 'DD-MM-YYYY HH:mm');
}

export function mapStateToProps(state: Object): Object {
  return {
    coreLimit: get('rateLimit.core.limit', state),
    searchLimit: get('rateLimit.search.limit', state),
    searchRemaining: get('rateLimit.search.remaining', state),
    searchReset: formatDate(get('rateLimit.search.reset', state)),
  };
}

export default connect(mapStateToProps);
