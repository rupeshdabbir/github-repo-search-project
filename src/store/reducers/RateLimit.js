import assignAll from 'lodash/fp/assignAll';

const initialState = {
  core: {},
  search: {},
};

export default function rateLimitReducer(state: Object = initialState, action: Object) {
  switch (action.type) {

    case 'API_RATE_LIMIT_SUCCESS':
      return assignAll([
        state,
        {
          core: action.payload.core,
          search: action.payload.search,
        },
      ]);

    case 'API_RATE_LIMIT_FAILURE': {
      if (!action.error) {return state;}
      return assignAll([
        state,
        {
          payload: action.payload,
        },
      ]);
    }

    default:
      return state;

  }
}
