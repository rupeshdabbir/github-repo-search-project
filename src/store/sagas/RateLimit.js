import { put, take, takeLatest, select, call } from 'redux-saga/effects';
import keys from 'lodash/fp/keys';
import map from 'lodash/fp/map';
import get from 'lodash/fp/get';
import every from 'lodash/fp/every';
import flow from 'lodash/fp/flow';
import castArray from 'lodash/fp/castArray';
import api from 'store/api';

const isCachedAction = flow(castArray, every(get('meta.fromCache')));

export function* getRateLimit(actions) {
  if (isCachedAction(actions)) {
    const latestRateLimit = yield select(get('cache.rateLimit.latest'));
    yield put({
      payload: latestRateLimit,
      type: 'API_RATE_LIMIT_SUCCESS',
    });
    return;
  }

  try {
    const rateLimit = yield call(api.get, 'rate_limit');
    yield put({
      payload: get('data.resources', rateLimit),
      type: 'API_RATE_LIMIT_SUCCESS',
    });
  } catch (err) {
    yield put({
      error: true,
      payload: err,
      type: 'API_RATE_LIMIT_FAILURE',
    });
  }
}

export function* watchGetRateLimit() {
  const types = ['REPOS_SUCCESS', 'FOLLOWERS_SUCCESS', 'PROFILE_SUCCESS'];

  yield [combineLatest(types, getRateLimit), takeLatest('API_RATE_LIMIT_REQUEST', getRateLimit)];
}

/**
 * Tracks an array of action types and only calls the saga
 * when they all fire.
 * Resets once all actions fire
 *
 * Modified from https://github.com/jhewlett/redux-saga-combine-latest
 */
function* combineLatest(types, saga) {
  let actions = {};
  const getActions = map((type) => actions[type]);

  while (true) {
    const action = yield take(types);
    actions[action.type] = action;
    if (keys(actions).length === types.length) {
      yield saga(getActions(types));
      actions = {};
    }
  }
}
