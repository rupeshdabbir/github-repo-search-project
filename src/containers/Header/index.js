// @flow

import React, { Component } from 'react';
import { StyleSheet } from 'aphrodite/no-important';
import Header from 'components/Header';
import Container from 'components/Container';
import connect from './connect';

type Props = {
  onSubmit: Function,
  searchQuery: string,
  searchTerm: string,
  searchForRepos: Function,
};

export class HeaderContainer extends Component {
  static defaultProps = {
    searchTerm: '',
  };

  constructor(props: Props) {
    super(props);
    this.handleSearchRepos(props.searchQuery);
  }

  componentWillReceiveProps(nextProps: Props) {
    if (nextProps.searchQuery !== this.props.searchQuery) {
      this.handleSearchRepos(nextProps.searchQuery);
    }
  }

  handleSearchRepos(searchQuery: string): void {
    if (!searchQuery) {
      return;
    }
    this.props.searchForRepos(searchQuery);
  }

  render() {
    const { onSubmit, searchTerm } = this.props;

    return (
      <Container rootStyle={styles.HeaderContainer}>
        <Header onSubmit={onSubmit} searchTerm={searchTerm} />
      </Container>
    );
  }
}

const styles = StyleSheet.create({
  HeaderContainer: {
    backgroundColor: '#f4f4f4',
    borderBottom: '1px solid #aaa',
    boxShadow: '0px 0px 2px 1px rgba(0, 0, 0, .2)',
  },
});

export default connect(HeaderContainer);
