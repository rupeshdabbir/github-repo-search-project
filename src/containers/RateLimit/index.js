import React, { Component } from 'react';
import { css, StyleSheet } from 'aphrodite/no-important';
import 'suitcss-utils-flex/lib/flex.css';
import RateLimit from 'components/RateLimit';
import isUndefined from 'lodash/fp/isUndefined';
import connect from './connect';

type Props = {
  coreLimit: ?number,
  dispatch: Function,
  searchLimit: ?number,
  searchRemaining: ?number,
  searchReset: ?string,
};

export class RateLimitContainer extends Component {
  constructor(props: Props) {
    super(props);
    const { coreLimit, searchLimit, dispatch } = props;

    if (isUndefined(coreLimit) || isUndefined(searchLimit)) {
      dispatch({ type: 'API_RATE_LIMIT_REQUEST' });
    }
  }

  render() {
    const { searchLimit, searchRemaining, searchReset } = this.props;

    return (
      <div className="u-flex u-flexAlignItemsCenter">
        <p className={css(styles.RateLimitContainer_label)}>API requests</p>
        <div className={`${css(styles.RateLimitContainer)} u-flex`}>
          <div className={css(styles.RateLimitContainer_item)}>
            <RateLimit label="Search" remaining={searchRemaining} limit={searchLimit} reset={searchReset} />
          </div>
        </div>
      </div>
    );
  }
}

const styles = StyleSheet.create({
  RateLimitContainer: {
    backgroundColor: '#333',
    color: '#eee',
    padding: '5px 8px',
    borderRadius: 5,
  },

  RateLimitContainer_label: {
    fontSize: 14,
    marginRight: 10,
  },

  RateLimitContainer_item: {
    ':first-child': {
      marginRight: 5,
    },

    ':last-child': {
      marginLeft: 5,
    },
  },
});

export default connect(RateLimitContainer);
