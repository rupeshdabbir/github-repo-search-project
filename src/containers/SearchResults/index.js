// @flow

import React from 'react';
import { css, StyleSheet } from 'aphrodite/no-important';
import SearchTable from 'components/SearchTable';
import Pagination from 'components/Pagination';
import ResultsMessage from 'components/ResultsMessage';
import ErrorBoundary from 'components/ErrorBoundary';
import Loading from 'components/Loading';
import Container from 'components/Container';
import connect from './connect';

type Props = {
  isPending: boolean,
  pagination: ?Object,
  searchTerm: string,
  totalResults: number,
  userIds: Array<number>,
};

export function SearchResultsContainer(props: Props) {
  const { isPending, pagination, searchTerm, totalResults, userIds, totalResponse } = props;

  console.log('props', props);

  if (!searchTerm) {
    return null;
  }

  if (isPending) {
    return (
      <div className={css(styles.SearchResultsContainer_wrapLoading)}>
        <Loading />
      </div>
    );
  }

  return (
    <Container>
      <div className={css(styles.SearchResultsContainer_item, styles.SearchResultsContainer_wrapResultsMessage)}>
        <ErrorBoundary>
          <ResultsMessage searchTerm={searchTerm} resultsTotal={totalResults} pageTotal={userIds.length} />
        </ErrorBoundary>
      </div>
      <div className={css(styles.SearchResultsContainer_item)}>
        <ErrorBoundary>
          {/* <SearchResults searchTerm={searchTerm} ids={userIds} entities={userEntities} total={totalResults} /> */}
          <SearchTable totalResponse={totalResponse} />
        </ErrorBoundary>
      </div>
      <div className={css(styles.SearchResultsContainer_item)}>
        <ErrorBoundary>
          <Pagination {...pagination} />
        </ErrorBoundary>
      </div>
    </Container>
  );
}

SearchResultsContainer.defaultProps = {
  pagination: {},
  searchTerm: '',
};

const styles = StyleSheet.create({
  SearchResultsContainer_wrapResultsMessage: {
    marginBottom: 30,
  },

  SearchResultsContainer_wrapLoading: {
    marginTop: 25,
  },

  SearchResultsContainer_item: {
    marginTop: 15,
  },
});

export default connect(SearchResultsContainer);
