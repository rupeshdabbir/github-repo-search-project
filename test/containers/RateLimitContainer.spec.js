import React from 'react';
import renderer from 'react-test-renderer';
import {StyleSheetTestUtils} from 'aphrodite';

import {RateLimitContainer} from 'containers/RateLimit';

describe('Container: RateLimitContainer', () => {

  beforeAll(() => {
    StyleSheetTestUtils.suppressStyleInjection();
  });

  afterAll(() => {
    StyleSheetTestUtils.clearBufferAndResumeStyleInjection();
  });

  it('should render', () => {
    const component = renderer.create(
      <RateLimitContainer
        coreLimit={100}
        coreRemaining={100}
        coreReset="some date"
        searchLimit={100}
        searchRemaining={100}
        dispatch={jest.fn()}
        searchReset="some date"
      />
    );
    expect(component).toMatchSnapshot();
  });

});
