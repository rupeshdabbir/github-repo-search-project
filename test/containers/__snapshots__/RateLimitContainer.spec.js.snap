// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Container: RateLimitContainer should render 1`] = `
<div
  className="u-flex u-flexAlignItemsCenter"
>
  <p
    className="RateLimitContainer_label_18lea6b"
  >
    API requests
  </p>
  <div
    className="RateLimitContainer_xe3jre u-flex"
  >
    <div
      className="RateLimitContainer_item_ygrcmc"
    >
      <p
        className="RateLimit_f9e8qy u-flex"
        title="Limit resets at some date"
      >
        <span
          className="RateLimit_label_1rcbo1g"
        >
          Search
          :
        </span>
        <span
          className="RateLimit_high_9hrz9f"
        >
          100
        </span>
      </p>
    </div>
  </div>
</div>
`;
